packages/expect/src/__tests__/__snapshots__/matchers.test.js.snap

```diff
@@ -3002,451 +2980,347 @@ Difference:
+
+Received value: <red>1</>"
 `;
 
 exports[`.toHaveProperty() {pass: true} expect({"a": {"b": {"c": {"d": 1}}}}).toHaveProperty('a.b.c.d', 1) 1`] = `
-"<dim>expect(</><red>object</><dim>).</>not<dim>.toHaveProperty(</><green>path</><dim>, </><green>value</><dim>)</>
-
-Expected the object:
-  <red>{\\"a\\": {\\"b\\": {\\"c\\": {\\"d\\": 1}}}}</>
-Not to have a nested property:
-  <green>\\"a.b.c.d\\"</>
-With a value of:
-  <green>1</>
-"
+"<dim>expect(</><red>received</><dim>).</>not<dim>.</>toHaveProperty<dim>(</><green>path</><dim>, </><green>value</><dim>)</>
+
+Expected path: <green>\\"a.b.c.d\\"</>
+
+Expected value: not <green>1</>"
 `;
 
 exports[`.toHaveProperty() {pass: true} expect({"a": {"b": {"c": 5}}}).toHaveProperty('a.b', {"c": 5}) 1`] = `
-"<dim>expect(</><red>object</><dim>).</>not<dim>.toHaveProperty(</><green>path</><dim>, </><green>value</><dim>)</>
-
-Expected the object:
-  <red>{\\"a\\": {\\"b\\": {\\"c\\": 5}}}</>
-Not to have a nested property:
-  <green>\\"a.b\\"</>
-With a value of:
-  <green>{\\"c\\": 5}</>
-"
+"<dim>expect(</><red>received</><dim>).</>not<dim>.</>toHaveProperty<dim>(</><green>path</><dim>, </><green>value</><dim>)</>
+
+Expected path: <green>\\"a.b\\"</>
+
+Expected value: not <green>{\\"c\\": 5}</>"
 `;
 
 exports[`.toHaveProperty() {pass: true} expect({"a": {"b": undefined}}).toHaveProperty('a.b') 1`] = `
-"<dim>expect(</><red>object</><dim>).</>not<dim>.toHaveProperty(</><green>path</><dim>)</>
+"<dim>expect(</><red>received</><dim>).</>not<dim>.</>toHaveProperty<dim>(</><green>path</><dim>)</>
 
-Expected the object:
-  <red>{\\"a\\": {\\"b\\": undefined}}</>
-Not to have a nested property:
-  <green>\\"a.b\\"</>
-"
+Expected path: not <green>\\"a.b\\"</>
+
+Received value: <red>undefined</>"
 `;
 
 exports[`.toHaveProperty() {pass: true} expect({"a": {"b": undefined}}).toHaveProperty('a.b', undefined) 1`] = `
-"<dim>expect(</><red>object</><dim>).</>not<dim>.toHaveProperty(</><green>path</><dim>, </><green>value</><dim>)</>
-
-Expected the object:
-  <red>{\\"a\\": {\\"b\\": undefined}}</>
-Not to have a nested property:
-  <green>\\"a.b\\"</>
-With a value of:
-  <green>undefined</>
-"
+"<dim>expect(</><red>received</><dim>).</>not<dim>.</>toHaveProperty<dim>(</><green>path</><dim>, </><green>value</><dim>)</>
+
+Expected path: <green>\\"a.b\\"</>
+
+Expected value: not <green>undefined</>"
+`;
+
+exports[`.toHaveProperty() {pass: true} expect({"a": {}}).toHaveProperty('a.b', undefined) 1`] = `
+"<dim>expect(</><red>received</><dim>).</>not<dim>.</>toHaveProperty<dim>(</><green>path</><dim>, </><green>value</><dim>)</>
+
+Expected path: <green>\\"a.b\\"</>
+Received path: <red>\\"a\\"</>
+
+Expected value: not <green>undefined</>
+Received value:     <red>{}</>
+
+<dim>Because a positive assertion passes for expected value undefined if the property does not exist, therefore this negative assertion fails unless the property does exist and has a defined value</>"
```

Comment:
I love this

```suggestion
<dim>Because a positive assertion passes for expected value undefined if the property does not exist, this negative assertion fails unless the property does exist and has a defined value</>"
```

Reply:
Or make it clear that this will be fixed?
`.not.toHaveProperty(path, undefined) should not throw if the property does not exist. This will be fixed in the next major version.`

Reply:
@jeysal Good question. The improved report for this edge case seems stupid without some explanation, but I hope this double-meaning of `undefined` in a negative assertion is very rare.

I will think how to explain the breaking change when the time comes.

To minimize upgrade friction so people get benefit of performance and circus in Jest 25, let’s limit breaking assertion changes to #8023 and aim for a bigger batch in Jest 26 or 27.